### PR TITLE
Ag/rebuild/full app rebuild

### PR DESCRIPTION
**Pull request by andriy1507**
Used FusedLocationProvider for geolocation.

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -4,9 +4,9 @@
     xmlns:tools="http://schemas.android.com/tools"
     package="com.goryachok.forecastapp">
 
-    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
-    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
-    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
```
Comment (EugeneKryvoshey) [inline app/src/main/AndroidManifest.xml]: Please revert that. Export and use only 1 style settings file.

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -17,9 +17,10 @@
         android:icon="@mipmap/ic_launcher"
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
-        android:supportsRtl="true"
         android:screenOrientation="portrait"
+        android:supportsRtl="true"
```
Comment (EugeneKryvoshey) [inline app/src/main/AndroidManifest.xml]: Please revert this, app don't has arabic/hebrew translations.

File: app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import androidx.core.location.LocationManagerCompat
+import com.google.android.gms.location.LocationCallback
+import com.google.android.gms.location.LocationRequest
+import com.google.android.gms.location.LocationResult
+import com.google.android.gms.location.LocationServices
+
+abstract class LocationProvider(private val context: Context) {
+
+    companion object {
+        private val MOCK_LOCATION =
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt]: Please rename to "Default_LOCATION".

File: app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            Location("MOCK PROVIDER").apply { latitude = 90.0; longitude = 0.0 }
+
+        private var currentLocation: Location? = null
+    }
+
+    private val locationProviderClient = LocationServices.getFusedLocationProviderClient(context)
+
+    private val locationRequest by lazy {
+        LocationRequest.create().apply {
+            priority = LocationRequest.PRIORITY_LOW_POWER; numUpdates = 1; interval = 25
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt]: Please, extract magic numbers as constants.

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt [hunk 5]
```diff
@@ -4,41 +4,92 @@ import android.app.SearchManager
...
+
+    val viewModel by lazy {
+        ViewModelProvider(
+            this,
+            viewModelFactory
+        ).get(MainViewModel::class.java)
+    }
+
+    private val connectivityListener by lazy {
+        ConnectivityListener(
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Please refactor this, that ConnectivityListener has 2 callbacks but not show/hide snackbar functions.

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt [hunk 6]
```diff
@@ -4,41 +4,92 @@ import android.app.SearchManager
...
-                searchView.apply {
-                    clearFocus()
-                    setQuery("", false)
+                return if (connectivityListener.isNetworkAvailable) {
+                    searchView.apply {
+                        clearFocus()
+                        setQuery("", false)
+                    }
+                    (pagerAdapter.getItem(forecast_viewPager.currentItem) as? MyFragment)?.onSearchRequest(
+                        query ?: ""
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Please, change to String.orEmpty()

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt [hunk 7]
```diff
@@ -49,16 +100,46 @@ class MainActivity : AppCompatActivity() {
...
+                if (connectivityListener.isNetworkAvailable)
+                    passCoordinates()
             }
         }
         return super.onOptionsItemSelected(item)
     }
 
+    override fun onPause() {
+        super.onPause()
+        connectivityListener.isFirstLaunch = true
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Please remove this.

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt [hunk 8]
```diff
@@ -49,16 +100,46 @@ class MainActivity : AppCompatActivity() {
...
+                    location
+                )
+                viewModel.locationCache = location
+                viewModel.requestCache = ""
+            }
+            .addOnFailureListener {
+                AlertDialog.Builder(this)
+                    .setTitle("Error")
+                    .setMessage(it.message)
+                    .setPositiveButton("Close") { dialog, _ ->
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Please, extract to string.xml

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt [hunk 9]
```diff
@@ -49,16 +100,46 @@ class MainActivity : AppCompatActivity() {
...
+            .addOnSuccessListener { location ->
+                (pagerAdapter.getItem(forecast_viewPager.currentItem) as? MyFragment)?.onLocationRequest(
+                    location
+                )
+                viewModel.locationCache = location
+                viewModel.requestCache = ""
+            }
+            .addOnFailureListener {
+                AlertDialog.Builder(this)
+                    .setTitle("Error")
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Please, extract to string.xml

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                dailyForecast_cityName_textView.text =
+                    getString(R.string.location_template, it.city.name, it.city.country)
+                adapter = DailyForecastAdapter().apply {
+                    setNewItemList(it.list)
+                }
+                dailyForecast_recyclerView.adapter = adapter
+
+                dailyForecast_recyclerView.apply {
+                    layoutManager = LinearLayoutManager(this@DailyFragment.context)
+                }
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt]: Please refactor this.

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+                dailyLoading_progressBar.visibility = ProgressBar.INVISIBLE
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt]: Please, use View.GONE if possible.

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+            })
+        }
+        viewModel.loadData.observe(viewLifecycleOwner, Observer {
+            dailyLoading_progressBar.visibility = ProgressBar.VISIBLE
+        })
+        viewModel.errorData.observe(viewLifecycleOwner, Observer { result ->
+            this.context?.let { _ ->
+                dailyLoading_progressBar.visibility = ProgressBar.INVISIBLE
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt]: Please, use View.GONE if possible.

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt [hunk 13]
```diff
@@ -52,23 +45,28 @@ class HourlyFragment : MyFragment() {
                     layoutManager = LinearLayoutManager(this@HourlyFragment.context)
                 }
                 hourlyForecast_recyclerView.adapter = adapter
+                hourlyLoadingProgressBar.visibility = ProgressBar.INVISIBLE
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt]: Please, use View.GONE if possible.

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt [hunk 14]
```diff
@@ -52,23 +45,28 @@ class HourlyFragment : MyFragment() {
...
-        }
-        viewModel.errorData.observe(viewLifecycleOwner, Observer {
-            AlertDialog.Builder(this.context)
-                .setTitle("Error")
-                .setMessage(it.exception.message)
-                .setPositiveButton("Close") { dialog, _ ->
-                    dialog.dismiss()
+            viewModel.errorData.observe(viewLifecycleOwner, Observer { result ->
+                this.context?.let { _ ->
+                    hourlyLoadingProgressBar.visibility = ProgressBar.INVISIBLE
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt]: Please, use View.GONE if possible.

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt [hunk 15]
```diff
@@ -1,10 +1,31 @@
...
 
     abstract fun onSearchRequest(request: String)
 
-    abstract fun onLocationRequest()
+    abstract fun onLocationRequest(loc: Location?)
+
+    override fun onResume() {
+        super.onResume()
+        val mainViewModel = activity?.let { (it as MainActivity).viewModel }
+        if (mainViewModel?.requestCache != "") {
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Please change to String.isNotEmpty()

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt [hunk 16]
```diff
@@ -1,10 +1,31 @@
...
+            onSearchRequest(mainViewModel?.requestCache ?: "")
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Please change to String.orEmpty()

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt [hunk 17]
```diff
@@ -1,10 +1,31 @@
...
+        } else {
+            onLocationRequest(mainViewModel.locationCache)
+        }
+    }
+
+    protected val errorDialog: AlertDialog by lazy {
+        AlertDialog.Builder(this.requireContext()).setTitle("Error")
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Please, extract to string.xml

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt [hunk 18]
```diff
@@ -1,10 +1,31 @@
...
+            .setPositiveButton("Close") { dialog, _ -> dialog.dismiss() }.create()
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Please, extract to string.xml

File: app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DailyForecastViewHolder {
+        val itemView = LayoutInflater.from(parent.context)
+            .inflate(R.layout.forecast_item_layout, parent, false)
+        return DailyForecastViewHolder(
+            itemView
+        )
+    }
+
+    override fun getItemCount(): Int {
+        return forecast.size
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt]: Please, inline this.

File: app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+        forecast[position].let { holder.bind(it) }
+    }
+
+    fun setNewItemList(forecast: List<Forecast>) {
+        this.forecast = forecast.toMutableList()
+        notifyItemRangeChanged(0, this.forecast.size - 1)
+    }
+
+    class DailyForecastViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        var temp: TextView = itemView.findViewById(R.id.temp_textView)
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt]: Please, remove findViewById.

File: app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+        CoroutineScope(Dispatchers.IO).launch {
+            _loadData.postValue(Result.Loading)
+            when (val result: Result<ForecastEntity> = repository.getDataByCity(city)) {
+                is Result.Success -> {
+                    val newList: List<Forecast> =
+                        result.data.list.getEveryEighth()
+                    withContext(Dispatchers.Main) {
+                        _currentData.postValue(result.data.copy(list = newList))
+                    }
+//                    }
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt]: Please remove this.

File: app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+            when (val result: Result<ForecastEntity> = repository.getDataByCoordinates(
+                loc.latitude.toFloat(),
+                loc.longitude.toFloat()
+            )) {
+                is Result.Success -> {
+                    val newList = result.data.list.getEveryEighth()
+                    withContext(Dispatchers.Main) {
+                        _currentData.postValue(result.data.copy(list = newList))
+                    }
+//                    }
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt]: Please remove this.

File: app/src/main/res/layout/current_weather_fragment.xml [hunk 23]
```diff
@@ -86,22 +86,15 @@
...
-    <LinearLayout
-        android:layout_width="match_parent"
+    <ProgressBar
+        android:id="@+id/currentLoadingProgressBar"
+        style="?android:attr/progressBarStyle"
+        android:layout_width="wrap_content"
         android:layout_height="wrap_content"
-        android:layout_marginBottom="25dp"
-        android:orientation="vertical"
+        android:visibility="invisible"
```
Comment (EugeneKryvoshey) [inline app/src/main/res/layout/current_weather_fragment.xml]: Please, use View.GONE if possible.

File: app/src/main/res/layout/hourly_forecast_fragment.xml [hunk 24]
```diff
@@ -39,4 +39,14 @@
         app:layout_constraintStart_toStartOf="parent"
         app:layout_constraintTop_toBottomOf="@+id/hourlyForecast_textView" />
 
+    <ProgressBar
+        android:id="@+id/hourlyLoadingProgressBar"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:visibility="invisible"
```
Comment (EugeneKryvoshey) [inline app/src/main/res/layout/hourly_forecast_fragment.xml]: Please, use View.GONE if possible.

File: dependencies.gradle [hunk 25]
```diff
@@ -37,7 +43,7 @@ ext.kotlinVersion = "1.3.61"
 
 //SDK
 ext.toolsVer = "29.0.2"
-ext.minimalSdkVer = 15
+ext.minimalSdkVer = 23
```
Comment (EugeneKryvoshey) [inline dependencies.gradle]: Please revert this.

File: app/src/main/java/com/goryachok/forecastapp/ConnectivityListener.kt [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+import android.graphics.Color
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest
+import android.view.View
+import com.google.android.material.snackbar.Snackbar
+
+class ConnectivityListener(context: Context, layout: View) {
+    // TODO Move constants to separate file
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/ConnectivityListener.kt]: Please, just do it.

**Review by EugeneKryvoshey — CHANGES_REQUESTED** [review]

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (andriy1507) [reply app/src/main/AndroidManifest.xml]: Ok

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (andriy1507) [reply app/src/main/AndroidManifest.xml]: Done

File: app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt (hunk 3 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt (hunk 4 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/LocationProvider.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt (hunk 22 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt (hunk 5 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt (hunk 6 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt (hunk 7 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt (hunk 8 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt (hunk 9 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/activity/MainActivity.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt (hunk 10 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt (hunk 11 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt (hunk 12 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/DailyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt (hunk 13 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt (hunk 14 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/HourlyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt (hunk 15 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Changed to:
        if (!mainViewModel?.requestCache.isNullOrEmpty()) {

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt (hunk 16 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt (hunk 17 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt (hunk 18 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/fragment/MyFragment.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt (hunk 19 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt (hunk 20 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/view/recyclerviewadapters/DailyForecastAdapter.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt (hunk 21 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/viewmodel/DailyViewModel.kt]: Done

File: app/src/main/res/layout/current_weather_fragment.xml (hunk 23 above)
Comment (andriy1507) [reply app/src/main/res/layout/current_weather_fragment.xml]: Done

File: app/src/main/res/layout/hourly_forecast_fragment.xml (hunk 24 above)
Comment (andriy1507) [reply app/src/main/res/layout/hourly_forecast_fragment.xml]: Done

File: dependencies.gradle (hunk 25 above)
Comment (andriy1507) [reply dependencies.gradle]: Done

**Review by EugeneKryvoshey — APPROVED** [review]

**Pull request merged**